src/runtime.rs

```diff
@@ -280,113 +247,151 @@ impl Runtime {
     ///
     /// This is mostly useful to control the durability level for [on-demand inputs](https://salsa-rs.github.io/salsa/common_patterns/on_demand_inputs.html).
     pub fn report_synthetic_read(&self, durability: Durability) {
+        let changed_at = self.last_changed_revision(durability);
         self.local_state
-            .report_synthetic_read(durability, self.current_revision());
+            .report_synthetic_read(durability, changed_at);
     }
 
-    /// Obviously, this should be user configurable at some point.
-    pub(crate) fn report_unexpected_cycle(
+    fn create_cycle_error(
         &self,
+        db: &dyn Database,
+        mut dg: MutexGuard<'_, DependencyGraph>,
         database_key_index: DatabaseKeyIndex,
-        error: CycleDetected,
-        changed_at: Revision,
-    ) -> crate::CycleError<DatabaseKeyIndex> {
-        debug!(
-            "report_unexpected_cycle(database_key={:?})",
-            database_key_index
-        );
-
-        let mut query_stack = self.local_state.borrow_query_stack_mut();
-
-        if error.from == error.to {
-            // All queries in the cycle is local
-            let start_index = query_stack
-                .iter()
-                .rposition(|active_query| active_query.database_key_index == database_key_index)
-                .unwrap();
-            let mut cycle = Vec::new();
-            let cycle_participants = &mut query_stack[start_index..];
-            for active_query in &mut *cycle_participants {
-                cycle.push(active_query.database_key_index);
-            }
-
-            assert!(!cycle.is_empty());
-
-            for active_query in cycle_participants {
-                active_query.cycle = cycle.clone();
-            }
-
-            crate::CycleError {
-                cycle,
-                changed_at,
-                durability: Durability::MAX,
-            }
-        } else {
-            // Part of the cycle is on another thread so we need to lock and inspect the shared
-            // state
-            let dependency_graph = self.shared_state.dependency_graph.lock();
-
-            let mut cycle = Vec::new();
-            dependency_graph.push_cycle_path(
+        to_id: RuntimeId,
+    ) -> CycleDetected {
+        debug!("create_cycle_error(database_key={:?})", database_key_index);
+
+        let mut from_stack = self.local_state.take_query_stack();
+        let from_id = self.id();
+
+        // Extract the changed_at and durability values from the top of the stack;
+        // these reflect the queries which were executed thus far and which
+        // led to the cycle.
+        let changed_at = from_stack.last().unwrap().changed_at;
+        let durability = from_stack.last().unwrap().durability;
```

Comment:
Not sure this logic is correct -- here we look at the top of our stack, but what if:

* we have different values on different stacks
* we have a different value on *our* stack, which is before the last element, and still participates in a cycle

I think we should compute that as minimum across all queries in `for_each_cycle_participant`.

Reply:
This code was removed in the latest version. That said, there is similar logic in the places where we catch cycle panics, and when you asked earlier I got to having similar doubts! We could certainly detect that information. I'd like to write a test that exposes the problem first.